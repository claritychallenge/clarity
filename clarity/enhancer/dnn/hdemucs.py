import torch
import torchaudio
from torchaudio.pipelines import HDEMUCS_HIGH_MUSDB_PLUS
from torchaudio.transforms import Fade


def apply_separation(mix, model, sample_rate, segment=10.0, overlap=0.1, device=None):
    """Separate audio into sources using the model.
    Args:
        mix (torch.Tensor): Audio tensor of shape (batch, channels, time).
        model (torch.nn.Module): Model to separate audio.
        sample_rate (int): Sample rate of audio.
        segment (float): Length of audio segments in seconds.
        overlap (float): Overlap between segments in seconds.
        device (torch.device): Device to use for inference.
    Returns:
        torch.Tensor: Separated audio tensor of shape (batch, sources, channels, time).
    """
    # Ensure shape (batch, channels, length)
    if mix.ndim == 1:
        # one track and mono audio
        mix = mix.unsqueeze(0)
    elif mix.ndim == 2:
        # one track and stereo audio
        mix = mix.unsqueeze(0)

    # If track is mono, repeat it to match the number of channels
    if mix.shape[1] == 1:
        mix = mix.repeat(1, 2, 1)

    batch, channels, length = mix.shape

    # Move model and audio to device
    mix = mix.to(device)
    model = model.to(device)

    # Normalize audio
    mix = mix / torch.max(torch.abs(mix))

    # Split audio into segments
    chunk_len = int(sample_rate * segment * (1 + overlap))
    start = 0
    end = chunk_len
    overlap_frames = overlap * sample_rate
    fade = Fade(fade_in_len=0, fade_out_len=int(overlap_frames), fade_shape="linear")

    separated_audio = torch.zeros(
        batch, len(model.sources), channels, length, device=device
    )

    while start < length - overlap_frames:
        chunk = mix[:, :, start:end]
        with torch.no_grad():
            out = model.forward(chunk)
        out = fade(out)
        separated_audio[:, :, :, start:end] += out
        if start == 0:
            fade.fade_in_len = int(overlap_frames)
            start += int(chunk_len - overlap_frames)
        else:
            start += chunk_len
        end += chunk_len
        if end >= length:
            fade.fade_out_len = 0

    # Move audio back to CPU
    separated_audio = separated_audio.cpu()

    return separated_audio


def separate_hdemucs(audio_track=None, segment=10.0, overlap=0.1, device=None):
    """Separate audio track using HDemucs.
    Args:
        audio_track (str): Paths of audio track.
        segment (float): Length of audio segments in seconds.
        overlap (float): Overlap between segments in seconds.
        device (torch.device): Device to use for inference.
    Returns:
        torch.Tensor: Separated audio tensor of shape (sources, channels, time).
    """
    # Load model
    bundle = HDEMUCS_HIGH_MUSDB_PLUS
    model = bundle.get_model()
    sources_list = model.sources

    if device is None:
        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    else:
        device = torch.device(device)

    model.to(device)

    # Load audio
    waveform, sample_rate = torchaudio.load(audio_track)

    sources = apply_separation(
        waveform, model, sample_rate, segment=segment, overlap=overlap, device=device
    ).squeeze(0)

    return dict(zip(sources_list, sources)), sample_rate
