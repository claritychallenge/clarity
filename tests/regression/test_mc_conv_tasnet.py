import numpy as np
import torch
import torchaudio
from omegaconf import OmegaConf
from tqdm import tqdm

from clarity.dataset.cec1_dataset import CEC1Dataset
from clarity.enhancer.dnn.mc_conv_tasnet import ConvTasNet, overlap_and_add


def test_overlap_add(regtest):
    signal = torch.Tensor(np.sin(np.arange(0, 2 * np.pi, 0.001) * 10))
    signal = torch.reshape(signal, (2, -1))

    frame_step = 100
    output = overlap_and_add(signal, frame_step, None)
    regtest.write(f"overlap add output: \n{output.detach().numpy()}\n")


def test_convtasnet(regtest):
    torch.manual_seed(0)
    cfg = {}
    cfg["path"] = {"exp_folder": "./"}
    cfg["mc_convtasnet"] = {
        "N_spec": 256,
        "N_spat": 128,  # 6 * 30
        "L": 20,
        "B": 256,
        "H": 512,
        "P": 3,
        "X": 6,
        "R": 4,
        "C": 1,  # num_speakers
        "num_channels": 6,  # should be consistent with dataloader num_channels
        "norm_type": "cLN",
        "causal": True,
        "mask_nonlinear": "relu",
        "device": "cpu",
    }

    cfg["test_dataset"] = {
        "scenes_folder": "tests/test_data/scenes",
        "scenes_file": "tests/test_data/metadata/scenes.test.json",
        "sr": 44100,
        "downsample_factor": 20,
        "wav_sample_len": None,
        "wav_silence_len": 0,
        "num_channels": 2,
        "norm": 1,
        "testing": False,
    }

    cfg["test_loader"] = {
        "batch_size": 1,
        "shuffle": False,
        "num_workers": 10,
    }
    cfg = OmegaConf.create(cfg)
    device = "cuda" if torch.cuda.is_available() else None
    test_set = CEC1Dataset(**cfg["test_dataset"])
    test_loader = torch.utils.data.DataLoader(dataset=test_set, **cfg["test_loader"])

    if cfg["test_dataset"]["downsample_factor"] != 1:
        down_sample = torchaudio.transforms.Resample(
            orig_freq=cfg.test_dataset["sr"],
            new_freq=cfg.test_dataset["sr"] // cfg.test_dataset["downsample_factor"],
            resampling_method="sinc_interpolation",
        )
        up_sample = torchaudio.transforms.Resample(
            orig_freq=cfg.test_dataset["sr"] // cfg.test_dataset["downsample_factor"],
            new_freq=cfg.test_dataset["sr"],
            resampling_method="sinc_interpolation",
        )

    with torch.no_grad():
        for batch in tqdm(test_loader, desc="testing"):
            noisy, scene = batch
            out = []
            for ear in ["left", "right"]:
                torch.cuda.empty_cache()
                # load denoising module
                den_model = ConvTasNet(**cfg.mc_convtasnet)

                den_model = torch.nn.parallel.DataParallel(den_model.to(device))
                den_model.eval()

                noisy = torch.reshape(noisy, (1, 6, -1))
                noisy = noisy.to(device).cpu()
                if cfg.test_dataset["downsample_factor"] != 1:
                    proc = down_sample(noisy)
                enhanced = (den_model(proc)).squeeze(1)
                enhanced = enhanced.cpu()
                if cfg.test_dataset["downsample_factor"] != 1:
                    enhanced = up_sample(enhanced)
                enhanced = torch.clamp(enhanced, -1, 1)
                out.append(enhanced.detach().cpu().numpy()[0])

            out = np.stack(out, axis=0).transpose()
    regtest.write(f"ctn output: \n{np.round(out, 6)}\n")
