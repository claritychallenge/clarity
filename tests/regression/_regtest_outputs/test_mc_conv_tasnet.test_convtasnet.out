ctn output:
[[ 0.001889 -0.008292]
 [ 0.001597 -0.008242]
 [ 0.001265 -0.008123]
 ...
 [-0.001017 -0.000657]
 [-0.000668 -0.000418]
 [-0.000345 -0.0002  ]]
